### ISSUE TITLE
Stage assignments remain after user's role is removed

### ISSUE THREAD

**Issue by NateWr** [pkp/pkp-lib#5396]
**Describe the bug**
A user assigned to a submission as an editor may later lose that role in the journal. When they do, they remain assigned to the submission as an editor even though they don't have that role.

The user's remaining roles allow them to access some submissions. This may happen if an editor or subeditor is reassigned as a reviewer, for example. This will lead to a circumstance in which the user can see that submission in their queue but can not access the workflow. It also means that they can access the submission details through the `/submissions` endpoint in the API, but not directly at `/submissions/<id>`. (See [comment](https://github.com/pkp/pkp-lib/issues/4868#issuecomment-574308149).)

**To Reproduce**
1. Assign an editor to a submission.
2. Change that editor's role so that they only have a reviewer role.
3. Visit the submissions list or the `/submissions` API endpoint.
4. The user will still have access to the submission details.

**What application are you using?**
 OJS/OMP `master` (3.2 pre-release) but I believe this effects most recent releases.

**Additional information**
As far as I can tell, this is a long-standing issue and does not lead to a problem in practice for a couple of reasons.

1. Editors probably hand off assigned submissions or see them through to completion before their role is revoked.
2. It's not common for editors to be assigned as an editor, then have their permissions revoked, then act as reviewers on the same submission. So in all but the rarest of cases no anonymous author data is getting revealed.

However, because stage assignments form the basis for access control, it would be a good idea to update stage assignments when user roles are changed. This presents a small conundrum.

If **stage assignments are removed**, a user will no longer see them in their Archive list. They'll lose a historical record of what they did edit at one time. Current editors will always be able to see the assignment by looking in the submission's activity log, but the user who is no longer assigned will not have access to this.

If **stage assignments are retained**, a user will still have access to some basic details about these submissions. Also, other editors may not know that this user is no longer truly assigned to the submission. They won't know that another user needs to be assigned.

I think that stage assignments should be removed when a user role is dropped. This will ensure that indicators in the UI (like submissions that are unassigned) will be updated accordingly.

Over time, submissions in the Archives will end up with no active assignments and once-editors who now have a different role (like reviewers) will no longer see a history of their editorial work in the Archives list. I think that's ok, but if it's not, then we should work backwards from this problem and find a way to display a user's historical editorial assignments to them in a different way.

**Comment by Vitaliy-1** [comment]
Just encountered this during 3.3 testing. The opposite is true also, I mean when a user was assigned to a submission in the role with lower permissions level, e.g., translator, and then promoted to the role with higher permissions, e.g., proofreader, the user won't have access to the assigned submission according to the new role; in this case - production ready files. It's still an uncommon scenario.

**Comment by NateWr** [comment]
> The opposite is true also

Yes, but I think that this is as it should be. If a user is assigned to a submission as a translator, granting them a proofreader role should not give them proofreader access to all submissions where they are assigned a a translator. Someone with two roles can be assigned in either or both roles, to limit their access to a specific submission.

> [me in original comment] I think that stage assignments should be removed when a user role is dropped.

I think I disagree with myself now. :joy: The historical record of who was assigned to what is important. Although this can be retrieved from the activity log, it seems like the assignments will be important for audits. For example, searching the archives for submissions assigned to an editor is an important tool when investigating allegations of misbehaviour.

I suggest we consider a `disabled` or `removed` flag on the stage assignments which would permit us to remove access to the assigned users but provide useful access to that information in the UI.

**Comment by ajnyga** [comment]
We should also consider section editor assignments to certain sections and categories. In a similar manner if you remove a section editor role, the assignment to specific sections/categories remain.

These are both mentioned here: https://github.com/pkp/pkp-lib/issues/5207

@NateWr maybe close that old issue and move anything usefull here?

**Comment by NateWr** [comment]
Good idea. There is some good discussion on that other issue. I'm copying over a few comments from #5207:

Idea from @asmecher for disabled stage assignments that doesn't require a flag in the database

> Wherever the stage assignment is used, add a join to the user_user_groups table (etc) to ensure that there is a live user group and not just a submission stage assignment. That would make a removal from a user group effective, but permit the restoration of the previous state if e.g. it was an accident or temporary change.

Comment from @alexxxmendonca 

> Giving the option to remove or retain existing assignments [when a role is removed] would be the best scenario but if that's too complicated to achieve, a simple confirmation such as "this action will remove any current section editor assignments this user might have, are you sure you want to proceed" or something along these lines.

Comment from @dihagan

> While we're at it: an option to reassign all that person's current assignments to another section editor.
